docs/design-proposals/2021-11-12-validation-webhook-for-CRs.md

```diff
@@ -0,0 +1,68 @@
+# Validation webhook for CRs
+
+## Summary
+
+This document presents a design to validate and admit webhook for CRs, which contains:
+
+- TidbCluster
+- TidbMonitor
+- TidbInitializer
+- Backup
+- Restore
+- BackupSchedule
+- TidbClusterAutoScaler
+
+### Goals
+
+Fix issues:
+
+- <https://github.com/pingcap/tidb-operator/issues/3949>
+- <https://github.com/pingcap/tidb-operator/issues/4065>
+- <https://github.com/pingcap/tidb-operator/issues/3468>
+- <https://github.com/pingcap/tidb-operator/issues/3213>
+- <https://github.com/pingcap/tidb-operator/issues/2534>
+- <https://github.com/pingcap/tidb-operator/issues/2490>
+
+### Non-Goals
+
+## Proposal
+
+- Add an admission controller named resource to admission server just like pod/statefulset/strategy.
+- Create resource package in pkg/webhook, the directory tree may like this:
+
+```Go
+pkg/webhook/resource/
+|- resource.go                        // resource register and resource interface 
+|- tidbcluster_resource.go            // validate or admit logic of TidbCluster
+|- tidbcluster_resource_test.go
+|- tidbmonitor_resource.go            // validate or admit logic of TidbMonitor
+|- tidbmonitor_resource_test.go 
+|- ...                                // other resources
+|- webhook.go            // resourceAdmissionControl struct, implementation interfaces in generic-admission-server and initialization of Resources.
+|- webhook_test.go        
+```
+
+- Interface `ValidateAdmitResource` implemented by every resource in `resource.go` may have the following methods:
+
+```Go
+type ValidateAdmitResource interface {
```

Comment:
Why not split into different interfaces for validation and mutation？

Reply:
You can also see interface design like [kubebuilder](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html)

Reply:
Ok, I will refer to it and optimize this proposal.